aten/src/ATen/native/LossNLL.cpp

```diff
@@ -0,0 +1,376 @@
+#include <ATen/ATen.h>
+#include <ATen/AccumulateType.h>
+#include <ATen/Dispatch.h>
+#include <ATen/Parallel.h>
+#include <ATen/TensorUtils.h>
+
+namespace at {
+namespace native {
+
+namespace {
+
+template <typename scalar_t>
+static void nll_loss_out_frame(
+    Tensor& output,
+    Tensor& total_weight,
+    const Tensor& input,
+    const Tensor& target,
+    const Tensor& weight,
+    int64_t reduction,
+    int64_t ignore_index) {
+  const auto n_dims = input.dim();
+  const auto n_classes = input.size(-1);
+
+  scalar_t* total_weight_data = total_weight.data_ptr<scalar_t>();
+  *total_weight_data = 0;
+
+  auto weight_contiguous = weight.contiguous();
+  const scalar_t* weight_data =
+      weight.defined() ? weight_contiguous.data_ptr<scalar_t>() : nullptr;
```

Comment:
If weight wasn't defined, wouldn't `weight.contiguous()` fail?

Reply:
It worked during my tests but I just had a look at the impl and I think I have to retest in debug mode. By default the `is_contiguous_` member of a TensorImpl (also UndefinedTensor) should be true ... contiguous() itself does ```  if (self.is_contiguous(memory_format)) return self; } ``` .. but there is a
```
#ifdef DEBUG
  AT_ASSERT(compute_contiguous() == is_contiguous_);
#endif
``` 
in `TensorImpl::is_contiguous()` ... and I fear the compute_contiguous() might throw. I will check. Thanks for bringing that up.

Reply:
I tested with DEBUG=1 build but apparently that does not define the DEBUG preprocessor symbol. From reviewing the source I believe that calling contiguous() on an undefined tensor breaks as soon as `DEBUG` is defined, therefore I added a helper function that checks if the tensor is defined before calling contiguous...